### PR TITLE
[WIP] Hook action and filter

### PR DESCRIPTION
**Pull request by morenol**
**Description:**

In this PR are we are adding the tools and functions to run the action hooks using a new module. Each action and filter should be configured with the HOOKS_EXTENSIONS. In that setting we need to indicate for each trigger which are the functions that needs to be run (and the order in which they are run) and if the tasks that we are running are sync or async

For instance:

```python
HOOKS_EXTENSIONS: {
      "preenrollment": {
               "pipeline": [
                     "somelibrary.filters.check_user_has_access_to_institution",
                     "other_library.filters.check_user_has_paid_the_course"
                     ],
               "async": false
     },
  
    "post_certificate_generated": {
               "pipeline": [
                     "some_library.actions.report_this_in_another_system",
                     "other_library.actions.create_badge_in_a_system"
                     ],
               "async": true
      }
}
```

They way to call them is, for actions:

```python
from edx_django_utils.hooks import do_action


do_action('post_certificate_generated', cert_generated)

```

For filters:

```python
from edx_django_utils.hooks import do_filter

course_list = do_filter('dashboard_course_list', 'course_list', course_list, user)

```

**Merge deadline:**

None

****Author concerns:**

I am not sure if this way to define if the function is async or not is fine. I think that for the template context related hooks we should use the already created tools within the plugin infrastructure.

### CODE REVIEW

**Comment by mariajgrimaldi** [comment]
If we assume that all actions are async and filters sync, can we consider this configuration:
```
HOOKS_EXTENSIONS: {
      "preenrollment": {
               "filters": [
                     "somelibrary.filters.check_user_has_access_to_institution",
                     "other_library.filters.check_user_has_paid_the_course"
                     ],
     },
  
    "post_certificate_generated": {
               "actions": [
                     "some_library.actions.report_this_in_another_system",
                     "other_library.actions.create_badge_in_a_system"
                     ],
      }
}
```

**Comment by mariajgrimaldi** [comment]
If we assume that all actions are async and filters sync, can we consider this configuration? 
```python
HOOKS_EXTENSIONS: {
      "preenrollment": {
               "filters": [
                     "somelibrary.filters.check_user_has_access_to_institution",
                     "other_library.filters.check_user_has_paid_the_course"
                     ],
     },
  
    "post_certificate_generated": {
               "actions": [
                     "some_library.actions.report_this_in_another_system",
                     "other_library.actions.create_badge_in_a_system"
                     ],
      }
}
```

**Comment by morenol** [comment]
I am not sure if that format or this one:

```python
HOOKS_EXTENSIONS: {
    "filters": {
       "preenrollment": [
                     "somelibrary.filters.check_user_has_access_to_institution",
                     "other_library.filters.check_user_has_paid_the_course"
                     ],
       },
    "actions": {
       "post_certificate_generated":  [
                     "some_library.actions.report_this_in_another_system",
                     "other_library.actions.create_badge_in_a_system"
                     ],
      }
}
```

Another option could be to just have a different django setting for the actions and for the filters.

File: edx_django_utils/hooks/filters.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+"""
+from logging import getLogger
+
+from django.conf import settings
+from .utils import get_cached_functions_for_hook
+
+log = getLogger(__name__)
+
+
+def do_filter(trigger_name, returned_key, *args, **kwargs):
```
Comment (mariajgrimaldi) [inline edx_django_utils/hooks/filters.py]: Should we consider when returned_key = None?

File: edx_django_utils/hooks/utils.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    Returns a tuple where the first item is the functions associated with the trigger and the second
+    item is a boolean that indicates wheter they should be run in an async way or not.
+
+    NOTE: These will be functions will be cached (in RAM not memcache) on this unique
+    combination. If we enable many new hooks to use this system, we may notice an
+    increase in memory usage as the entirety of these functions will be held in memory.
+    """
+    hook_config = _get_hook_configurations(trigger_name)
+
+    hook_is_async = hook_config.get("async", True)
```
Comment (mariajgrimaldi) [inline edx_django_utils/hooks/utils.py:18]: Maybe we could move this to _get_hook_configurations as is still part of the hook_config

File: edx_django_utils/hooks/filters.py (hunk 1 above)
Comment (morenol) [reply edx_django_utils/hooks/filters.py]: maybe this idea of returned_key is not ok, and the final dictionary is handled by the platform

**Pull request closed without merging**